### PR TITLE
feat(queryRules): add QueryRuleCustomData widget [part 2]

### PR DESCRIPTION
**Pull request by francoischalifour**
_This is the 2nd PR in the series "Merchandized Query Rules"._

## Summary

This creates the `QueryRuleCustomData` widget, responsible for showing custom data (`userData`) coming from Query Rules.

## Motivation

[Read RFC →](https://github.com/algolia/instantsearch-rfcs/blob/master/accepted/merchandized-query-rules.md)

### Basic

```jsx
<QueryRuleCustomData>
  {({ items }) =>
    items.map(({ banner, title, link }) => {
      if (!banner) {
        return null;
      }

      return (
        <section key={title}>
          <h2>{title}</h2>

          <a href={link}>
            <img src={banner} alt={title} />
          </a>
        </section>
      );
    })
  }
</QueryRuleCustomData>
```

### With default banner

```jsx
<QueryRuleCustomData
  transformItems={items => {
    if (items.length > 0) {
      return items;
    }

    return [
      {
        title: 'Kill Bill',
        banner: 'http://static.bobatv.net/IMovie/mv_2352/poster_2352.jpg',
        link: 'https://www.netflix.com/title/60031236',
      },
    ];
  }}
>
  {({ items }) =>
    items.map(({ banner, title, link }) => {
      if (!banner) {
        return null;
      }

      return (
        <section key={title}>
          <h2>{title}</h2>

          <a href={link}>
            <img src={banner} alt={title} />
          </a>
        </section>
      );
    })
  }
</QueryRuleCustomData>
```

## Stories

[See stories →](https://deploy-preview-2212--react-instantsearch.netlify.com/storybook/?selectedKind=QueryRuleCustomData&selectedStory=default)

## Related

- React InstantSearch
  - [`connectQueryRules`](https://github.com/algolia/react-instantsearch/pull/2210)
- InstantSearch.js
  - [`connectQueryRules`](https://github.com/algolia/instantsearch.js/pull/3597)
  - [`queryRuleCustomData`](https://github.com/algolia/instantsearch.js/pull/3600)
  - [`queryRuleContext`](https://github.com/algolia/instantsearch.js/pull/3602)
- [Merchandized Query Rules RFC](https://github.com/algolia/instantsearch-rfcs/blob/master/accepted/merchandized-query-rules.md)

### CODE REVIEW

File: packages/react-instantsearch-dom/src/components/QueryRuleCustomData.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  items: TItem[];
+};
+
+export type QueryRuleCustomDataProps<TItem> = {
+  items: TItem[];
+  className?: string;
+  children: (options: QueryRuleCustomDataRenderProps<TItem>) => React.ReactNode;
+};
+
+const QueryRuleCustomData: React.SFC<
```
Comment (samouss) [inline packages/react-instantsearch-dom/src/components/QueryRuleCustomData.tsx]:
```suggestion
const QueryRuleCustomData: React.FC<
```

File: packages/react-instantsearch-dom/src/widgets/QueryRuleCustomData.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { connectQueryRules, CustomUserData } from 'react-instantsearch-core';
+import PanelCallbackHandler from '../components/PanelCallbackHandler';
+import QueryRuleCustomData, {
+  QueryRuleCustomDataProps,
+} from '../components/QueryRuleCustomData';
+
+const QueryRuleCustomDataWidget: React.SFC<
```
Comment (samouss) [inline packages/react-instantsearch-dom/src/widgets/QueryRuleCustomData.tsx]:
```suggestion
const QueryRuleCustomDataWidget: React.FC<
```

File: packages/react-instantsearch-dom/src/components/QueryRuleCustomData.tsx (hunk 1 above)
Comment (Haroenv) [inline packages/react-instantsearch-dom/src/components/QueryRuleCustomData.tsx]: we use FC in the library so far

```suggestion
const QueryRuleCustomData: React.FC<
```

Comment (francoischalifour) [reply packages/react-instantsearch-dom/src/components/QueryRuleCustomData.tsx]: We use `SFC` for `Hits`: https://github.com/algolia/react-instantsearch/blob/642ba0b207deb8534c2d97bcde2f2b1fedd7e48c/packages/react-instantsearch-dom/src/components/Hits.tsx#L22

Apparently it's deprecated.

File: storybook/webpack.config.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ module.exports = {
             loader: 'babel-loader',
             options: {
               rootMode: 'upward',
+              presets: [['react-app', { typescript: true }]],
```
Comment (samouss) [inline storybook/webpack.config.js]: This line makes [the integration test fail](https://circleci.com/gh/algolia/react-instantsearch/17568). The `react-app` presets throw an error if `NODE_ENV` is `undefined`. The integration test does not explicitly set it so we have to update the command to set `NODE_ENV` to `development`. I've tried to catch the error at the script level but it seems that Storybook swallow it.

**Pull request merged**